Feat (Core): Implement abandoned job detection and recovery

jgambarios opened the pull request:
This pull request includes some refactoring and improvements to the job queue management system. 
The changes primarily focus on removing the job polling mechanism, enhancing job validation, and adding an abandoned job detector. Below are the most important changes:

### Removal of Job Polling Mechanism:

* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueConfig.java`](diffhunk://#diff-4c8b70e959cc00105adc8298bf8a977a3c18a6ceb5d55110bda6d023159a5e5aL13-L24): Removed the `pollJobUpdatesIntervalMilliseconds` parameter and its associated methods. [[1]](diffhunk://#diff-4c8b70e959cc00105adc8298bf8a977a3c18a6ceb5d55110bda6d023159a5e5aL13-L24) [[2]](diffhunk://#diff-4c8b70e959cc00105adc8298bf8a977a3c18a6ceb5d55110bda6d023159a5e5aL36-L44)
* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueConfigProducer.java`](diffhunk://#diff-1367d122323650421a37296f7a223becc19bb3d6c1d2a965dd52ae6bbd2f5e74L19-L23): Removed the default polling interval configuration and its usage. [[1]](diffhunk://#diff-1367d122323650421a37296f7a223becc19bb3d6c1d2a965dd52ae6bbd2f5e74L19-L23) [[2]](diffhunk://#diff-1367d122323650421a37296f7a223becc19bb3d6c1d2a965dd52ae6bbd2f5e74L33-R28)
* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java`](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L121-R122): Removed the polling scheduler and related methods. [[1]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L121-R122) [[2]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L477-L498)

### Addition of Abandoned Job Detector:

* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java`](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R112): Added `AbandonedJobDetector` as a dependency and integrated it into the job queue lifecycle methods. [[1]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R112) [[2]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L142-R151) [[3]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L170-L179) [[4]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L221-R209) [[5]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R230-L243)

### Enhancements to Job Validation:

* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java`](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R283-R303): Added validation for job parameters if the processor implements `Validator` and handled validation exceptions.

### Event Handling Improvements:

* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java`](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R283-R303): Replaced direct event firing with `JobUtil.sendEvents` for better event management. [[1]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R283-R303) [[2]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L543-R526) [[3]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L668-R649) [[4]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L721-R703) [[5]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L743-R724)

### Miscellaneous Changes:

* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java`](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L764-R743): Adjusted the job progress update interval from 2 to 3 seconds.

This PR fixes: #30367

jgambarios committed:
#30367 Deleted the `EventProducer` class as it is no longer needed. Added `JobEvent` interface to various job event classes and implemented validation logic in `ImportContentletsProcessor`. Introduced the `AbandonedJobDetector` class for detecting and handling abandoned jobs.

jgambarios committed:
#30367 Improved events handling in RealTimeJobMonitor and improvements in transaction handling on the JobQueueManagerAPIImpl

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-30367-Implement-Abandoned-Job-Detection-and-Recovery

jgambarios committed:
#30367 Improve error handling when creating a Job

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-30367-Implement-Abandoned-Job-Detection-and-Recovery

jgambarios committed:
#30367 Fixing IT

jgambarios committed:
#30367 Integration test for testing the abandoned job detection functionality.

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-30367-Implement-Abandoned-Job-Detection-and-Recovery

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-30367-Implement-Abandoned-Job-Detection-and-Recovery

jgambarios committed:
#30367 Refactor detectAndMarkAbandoned to return Optional<Job>

This change updates the detectAndMarkAbandoned method to return an Optional<Job> instead of null. This helps to avoid potential NullPointerExceptions and improves code readability. Corresponding updates were made to the affected classes and integration tests to handle the Optional return type appropriately.

jgambarios commented on dotCMS/src/main/java/com/dotcms/jobs/business/queue/PostgresJobQueue.java:
```diff
@@ -580,6 +618,60 @@ public Job nextJob() throws JobQueueDataException, JobLockingException {
+                return abandonedJob;
+            }
+
+            return null;
```
Done

jgambarios commented on dotCMS/src/main/java/com/dotcms/rest/api/v1/job/JobQueueResource.java:
```diff
@@ -54,37 +57,50 @@ public JobQueueResource(WebResource webResource, JobQueueHelper helper,
     @Consumes(MediaType.APPLICATION_JSON)
     @Produces(MediaType.APPLICATION_JSON)
-    public ResponseEntityView<String> createJob(
+    public Response createJob(
```
Now we can validate the job before creation, and if the job fails when validating, the exception needs to be handled:

```java
        } catch (JobValidationException e) {
            return ExceptionMapperUtil.createResponse(null, e.getMessage());
        }
```

The `ExceptionMapperUtil.createResponse` returns a `Response`, that's why I made the change.

jgambarios commented on dotCMS/src/main/java/com/dotcms/jobs/business/detector/AbandonedJobDetectorConfigProducer.java:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @return A new configuration object with the current property values
+     */
+    @Produces
```
Thanks, this is important to know.
Done.

jgambarios commented on dotCMS/src/main/java/com/dotcms/jobs/business/detector/AbandonedJobDetectorConfigProducer.java:
```diff
@@ -0,0 +1,48 @@
+ * </ul>
+ */
+@ApplicationScoped
+public class AbandonedJobDetectorConfigProducer {
```
Done.

jgambarios commented on dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java:
```diff
@@ -139,24 +136,19 @@ public class JobQueueManagerAPIImpl implements JobQueueManagerAPI {
             JobQueueConfig jobQueueConfig,
             CircuitBreaker circuitBreaker,
             @DefaultRetryStrategy RetryStrategy defaultRetryStrategy,
             RealTimeJobMonitor realTimeJobMonitor,
```
We can have multiple `RetryStrategy` implementation as each processor could have its own retry strategy, but if no retry strategy is provided for a processor we need to use the default, and the default retry strategy is produced in the `RetryStrategyProducer`.

nollymar commented on dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java:
```diff
@@ -218,7 +202,11 @@ public void start() {
+            // Start the abandoned job detector
+            abandonedJobDetector.start();
+
+            Logger.info(this,
```
is this going to log every 5 mins? couldn't it be logging too much?

jgambarios replied:
No, this is logged only when the `JobQueueManagerAPIImpl` starts, should be just once.

jgambarios committed:
#30367 Add default constructor for CDI proxy in AbandonedJobDetectorConfig

A protected no-arg constructor was added to AbandonedJobDetectorConfig to comply with CDI requirements. This ensures the class can be properly proxied and managed by the CDI container.

jgambarios committed:
#30367 Fixing unit tests

jgambarios committed:
Merge remote-tracking branch 'origin/main' into issue-30367-Implement-Abandoned-Job-Detection-and-Recovery
